src/robot/running/bodyrunner.py

```diff
@@ -556,3 +565,81 @@ def _run_finally(self, data, run):
                 return err
             else:
                 return None
+
+
+class WhileLimitExceeded(RuntimeError):
+    pass
+
+
+class WhileLimit:
+
+    @staticmethod
+    def parse_iteration_count(limit):
+        try:
+            limit = int(limit)
+            if limit <= 0:
+                raise ValueError
+            return limit
+        except ValueError:
+            raise ValueError("Iteration limit must be positive integer when using "
+                             f"'x' or 'times', got: '{limit}'")
+
+    @classmethod
+    def create(cls, limit, variables):
+        try:
+            if not limit:
+                return IterationCountLimit(100)
+            value = limit.replace('limit=', '')
+            if value.upper() == 'NONE':
+                return NoLimit()
+            value = variables.replace_scalar(value)
+            if value.endswith('x'):
```

Comment:
`value` isn't necessarily a string here so `endswith` isn't safe. Changing `replace_scalar` to `replace_string` ought to fix that but it would be good to add a test as well.